### PR TITLE
Allows passing the logger directly to the commands

### LINKED ISSUE #13: Allow passing a logger to commands
That would allow avoiding sth like this:

```js

				.then( () => {
					return execCommand.execute( execOptions( `git checkout ${ repository.branch }` ) );
				} )
				.then( ( response ) => {
					log.concat( response.logs );
				} )
				.then( () => {
					return execCommand.execute( execOptions( `git pull origin ${ repository.branch }` ) );
				} )
				.then( ( response ) => {
					log.concat( response.logs );
				} )
```

### PR DESCRIPTION
**Pull request by pomek**
```
Internal: Allows passing the logger directly to the commands. Closes #13.
```

---

Most probably this ticket will resolve the #45 but it needs to be verified.

### CODE REVIEW

File: lib/utils/getlogger.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+		error( msg ) {
+			if ( msg instanceof Error ) {
+				msg = msg.stack;
+			}
+
+			return logger._log( 'error', msg );
+		},
+
+		/**
+		 * @param {Logs} responseLogs
```
Comment (Reinmar) [inline lib/utils/getlogger.js]: Wrong variable name or method name or both. Why would logger have to know that what you pass it is response logs and not any logs? And `concat()` I've got some doubts whether `concat()` is the right name too, but I can't propose something better.

File: lib/utils/getlogger.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+		concat( responseLogs ) {
+			responseLogs.info.forEach( ( msg ) => logger.info( msg ) );
+
+			responseLogs.error.forEach( ( err ) => logger.error( err ) );
+		},
+
+		/**
+		 * @returns {Logs}
+		 */
+		all() {
```
Comment (Reinmar) [inline lib/utils/getlogger.js]: If it returns anything, the name should contain a verb.

File: lib/utils/getlogger.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+			return {
+				error: logsCollection.get( 'error' ),
+				info: logsCollection.get( 'info' ),
+			};
+		},
+
+		/**
+		 * @returns {String|undefined}
+		 */
+		getLastInserted() {
```
Comment (Reinmar) [inline lib/utils/getlogger.js]: You do not insert docs. It's an implementation detail what this util does with the logs.

File: lib/commands/update.js [hunk 4]
```diff
@@ -51,51 +50,39 @@ module.exports = {
...
-					resolve( { logs: log.all() } );
+				.then( () => {
+					return execCommand.execute( getExecData( `git pull origin ${ data.repository.branch }` ), logger );
 				} )
-				.catch( ( error ) => {
-					log.error( error );
-
-					reject( { logs: log.all() } );
+				.then( () => {
+					resolve( { logs: logger.all() } );
```
Comment (Reinmar) [inline lib/commands/update.js]: Can't commands return the logger itself? Perhaps you could get rid of the whole logger `all()` method and the logger could be a real logger with a method to flush the logs to the std output. Right now it's just a storage, not a logger.

**Comment by Reinmar** [comment]
This PR improves the piece of the logger management which I reported in the ticket, but now I want to see more ;>.

The logger managment is a significant percantage of the code. It should be straightforward and transparent, while every command needs to make sure to concat logs from other commands it triggers, return them back when fisnish its job, etc. Also, the logger itself isn't a logger, because it doesn't log anything to the console. The logging logic is in... `displaylog`. All this can be improved.

I commented one some of the issues, but there's much more that can be done. E.g. why not using a single logger instance everywhere (unless a separate instance is needed)?

Try to prototype some solution. Focus on reducing code size (by getting rid of so much logs managment from the commands) and enclosing the logging logic in the logger.

**Comment by pomek** [comment]
I assume you don't remember when I said - we cannot pass an object (_real_ object) through the threads. 

Threads communicate with each other using simple structures like strings, arrays and simple objects (`key: value`).

That's the reason why commands don't log anything. Commands collect the logs and send a response to a parent thread. The parent thread is able to display the logs (using `displaylog` module).

We can use _real_ logger but we will get mixed logs from many threads. The output won't be readable.

And one more thing - sometimes an output from commands is not important for the process. For example, from `mgit update` command point of view (I will copy part of the code):

- `execCommand.execute( getExecData( 'git status -s' ) )`
  We need to know an output for the process but we don't want to show the output to a user.
- `execCommand.execute( getExecData( 'git fetch' ), logger )`
  We want to show an output for the user.
- `execCommand.execute( getExecData( 'git checkout ' + data.repository.branch ), logger )`
  We want to show an output for the user.
- `execCommand.execute( getExecData( 'git branch' ) )`
  We need to know an output for the process but we don't want to show the output to the user.
- `execCommand.execute( getExecData( 'git pull origin ' + data.repository.branch ), logger )`
  We want to show an output for the user.

This PR will resolve your ticket. For refactoring the logger I propose to report another ticket.

File: lib/utils/getlogger.js (hunk 1 above)
Comment (pomek) [inline lib/utils/getlogger.js]: I have a feeling the `concat` method is not needed anymore.

File: lib/utils/getlogger.js (hunk 3 above)
Comment (pomek) [inline lib/utils/getlogger.js]: Hm, I don't understand.

File: lib/commands/update.js (hunk 4 above)
Comment (pomek) [inline lib/commands/update.js]: Hmm, commands would return the logger but `child-process` module needs to pass simple structures through the threads. (https://github.com/cksource/mgit2/pull/46#issuecomment-279807388)

File: lib/utils/getlogger.js (hunk 1 above)
Comment (pomek) [inline lib/utils/getlogger.js]: Ok, I removed the method.

File: lib/utils/child-process.js [hunk 5]
```diff
@@ -14,7 +14,7 @@ process.on( 'message', onMessage );
  * @param {Options} data.options The options object.
  */
 function onMessage( data ) {
-	const log = require( './log' )();
+	const log = require( './getlogger' )();
```
Comment (ma2ciek) [inline lib/utils/child-process.js]: getLogger

Comment (Reinmar) [inline lib/utils/child-process.js]: The variable name is ok. The module name is ok too. Note that the getLogger() function is immediately executed.

Comment (ma2ciek) [inline lib/utils/child-process.js]: Aw, you're right. Overlooked that.

**Comment by pomek** [comment]
This PR is quite outdated. 

It's hard to resolve conflicts if I even don't remember what I wanted to do here.

We shouldn't be waiting for merge **a half of year**. If we don't want to merge the PR, we should say it as soon as possible. Now I'm confused. Should we close this PR?

**Comment by Reinmar** [comment]
Sure, I'm sorry. But this PR wasn't read at any point. It required a lot of work still and we had more important things to work on. If it was close to being ready, I'd organise the work differently, but in such a situation other priorities won.

Let's leave this PR as a future reference for yet another solution to this problem. AFAIR, for me, the proposed changes were incomplete and this whole logs management should be written from scratch.

**Comment by pomek** [comment]
I'm closing this PR because we need to rethink how the logger should behave (https://github.com/cksource/mgit2/pull/46#issuecomment-319031097).

**Pull request closed without merging**